src/uu/split/src/split.rs

```diff
@@ -591,6 +627,123 @@ fn custom_write_all<T: Write>(
+) -> std::io::Result<u64>
+where
+    R: BufRead,
+{
+    // Set read limit to io_blksize if specified
+    // Otherwise to OPT_IO_BLKSIZE_MAX
+    let read_limit = io_blksize.unwrap_or(OPT_IO_BLKSIZE_MAX) as u64;
+
+    // Try to read into buffer up to a limit
+    let num_bytes = reader
+        .by_ref()
+        .take(read_limit)
+        .read_to_end(buf)
+        .map(|n| n as u64)?;
+
+    if num_bytes < read_limit {
+        // Finite file or STDIN stream that fits entirely
+        // into a buffer within the limit
+        // Note: files like /dev/null or similar,
+        // empty STDIN stream,
+        // and files with true file size 0
+        // will also fit here
+        Ok(num_bytes)
+    } else if input == "-" {
+        // STDIN stream that did not fit all content into a buffer
+        // Most likely continuous/infinite input stream
+        return Err(io::Error::new(
+            ErrorKind::Other,
+            format!("{}: cannot determine input size", input),
+        ));
+    } else {
+        // Could be that file size is larger than set limit
+        // or could be an "infinite" file like /dev/zero
+        // Attempt direct `seek()` for the end of a file
+        let mut tmp_fd = File::open(Path::new(input))?;
+        let end = tmp_fd.seek(SeekFrom::End(0))?;
+        if end > 0 {
+            Ok(end)
+        } else {
+            // Edge case of either "infinite" file (i.e. /dev/zero)
+            // or some other "special" non-standard file type
+            // Give up and return an error
+            // TODO It might be possible to do more here
+            // to address all possible file types and edge cases
+            return Err(io::Error::new(
+                ErrorKind::Other,
+                format!("{}: cannot determine file size", input),
+            ));
+        }
+    }
+}
+
+/// Get the size of the input file in bytes
+/// Used only for subset of `--number=CHUNKS` strategy, as there is a need
+/// to determine input file size upfront in order to know chunk size
+/// to be written into each of N files/chunks:
+/// * N       split into N files based on size of input
+/// * K/N     output Kth of N to stdout
+/// * l/N     split into N files without splitting lines/records
+/// * l/K/N   output Kth of N to stdout without splitting lines/records
+///
+/// For most files the size will be determined by `len()` function of [`std::fs::metadata`]
+/// However, for the STDIN input stream and files at /dev, /proc, /sys and similar locations
+/// we will need to read from the `reader` in an attempt to find out the input size
+/// In those cases the `buf` might end up with either partial or entire input
+fn get_input_size<R>(
+    input: &String,
+    reader: &mut R,
+    buf: &mut Vec<u8>,
+    io_blksize: &Option<usize>,
+) -> std::io::Result<u64>
+where
+    R: BufRead,
+{
+    if input == "-"
+        || input.starts_with("/dev/")
+        || input.starts_with("/proc/")
+        || input.starts_with("/sys/")
```

Comment:
This feels inherently brittle and I'm wondering whether there's a better way to do it. For example, we could have links to all of these filesystems and that should still work. Based on the syscalls they make, I don't think GNU has exceptions like, though I'm not entirely sure what they are doing instead.

Reply:
I think they might just try to get the length first and try this other method if that returns `0`? I'm not sure.

Reply:
in GNU they first go straight to reading it into a buffer and then compare that to what metadata/filesystem is reporting + do seek on the file if those disagree (as well as trying to copy things arround into temp file/buffer, etc) and some other edge cases. I tried to avoid reading file into a buffer first for everything, but could re-implement to closer follow GNU approach

Reply:
I think doing what GNU does is the right call! Nice investigative work! (As a precaution: remember not to look at the GNU source code)

Reply:
should be ready to go